codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyrust/generator/RustAwsSdkShimGenerator.java

```diff
@@ -0,0 +1,1739 @@
+        var structureShapeName = toSnakeCase(shape.getId().getName());
+        if (isDafnyOption) {
+          yield TokenTree.of(
+            """
+            match (*%s).as_ref() {
+                crate::r#_Wrappers_Compile::Option::Some { value } =>
+                    Some(crate::conversions::%s::from_dafny(value.clone())),
+                _ => None,
+            }
+            """.formatted(dafnyValue, structureShapeName)
+          );
+        } else {
+          TokenTree result = TokenTree.of(
+            """
+            crate::conversions::%s::from_dafny(%s.clone())
+            """.formatted(structureShapeName, dafnyValue)
+          );
+          if (isRustOption) {
+            result =
+              TokenTree.of(TokenTree.of("Some("), result, TokenTree.of(")"));
+          }
+          yield result;
+        }
+      }
+      default -> throw new IllegalArgumentException(
+        "Unsupported shape type: %s".formatted(shape.getType())
+      );
+    };
+  }
+
+  private TokenTree toDafnyVariantMemberForOperationRequest(
+    Shape parent,
+    MemberShape member
+  ) {
+    Shape targetShape = model.expectShape(member.getTarget());
+    String snakeCaseMemberName = toSnakeCase(member.getMemberName());
+    boolean isRequired = member.hasTrait(RequiredTrait.class);
+    // These rules were mostly reverse-engineered from inspection of Rust SDKs,
+    // and may not be complete!
+    boolean isRustRequired =
+      (isRequired &&
+        !operationIndex.isOutputStructure(parent) &&
+        !operationIndex.isInputStructure(parent) &&
+        !targetShape.isStructureShape()) ||
+      (operationIndex.isOutputStructure(parent) &&
+        targetShape.isIntegerShape());
+    return toDafny(
+      targetShape,
+      "value." + snakeCaseMemberName,
+      !isRustRequired,
+      !isRequired
+    );
+  }
+
+  /**
+   * @param isRustOption Whether the Rust type is an Option<...> or not.
+   *                     Rust's structures avoid Options when not strictly necessary depending on context.
+   * @param isDafnyOption Whether the Dafny type is an Option<...> or not.
+   *                      Dafny tends to be much more consistent about this.
+   *                      We generally trust that Dafny codegen aligns with the constraints,
+   *                      and hence is it safe to call unwrap() on Rust options when necessary.
+   */
+  private TokenTree toDafny(
+    final Shape shape,
+    final String rustValue,
+    boolean isRustOption,
+    boolean isDafnyOption
+  ) {
+    return switch (shape.getType()) {
+      case STRING -> {
+        if (shape.hasTrait(EnumTrait.class)) {
+          var enumShapeName = toSnakeCase(shape.toShapeId().getName());
+          if (isDafnyOption) {
+            yield TokenTree.of(
+              """
+              ::std::rc::Rc::new(match &%s {
+                  Some(x) => crate::_Wrappers_Compile::Option::Some { value: crate::conversions::%s::to_dafny(x.clone()) },
```

Comment:
It seems unfortunate that we `.clone()` so much data that's about to be converted anyway (so presumably the clone would never live long), though I suppose performance isn't the goal of this first pass.

Reply:
Yup, performance definitely isn't the highest priority, as long as we can come back and improve this later. I definitely found cloning everywhere was much simpler, to avoid any interference between the different branches when converting a structure for e.g.